waku/v2/node/waku_node.nim

```diff
@@ -409,80 +411,123 @@ proc mountRelay*(node: WakuNode,
+  info "mounting filter client"
+
+  node.wakuFilterClient = WakuFilterClient.new(node.peerManager, node.rng)
+  if node.started:
+    # Node has started already. Let's start filter too.
+    await node.wakuFilterClient.start()
+
+  node.switch.mount(node.wakuFilterClient, protocolMatcher(WakuFilterCodec))
+
+proc filterSubscribe*(node: WakuNode, pubsubTopic: PubsubTopic, contentTopics: ContentTopic|seq[ContentTopic], 
+                handler: FilterPushHandler, peer: RemotePeerInfo|string) {.async, gcsafe, raises: [Defect, ValueError].} =
   ## Registers for messages that match a specific filter. Triggers the handler whenever a message is received.
-  ## FilterHandler is a method that takes a MessagePush.
-  
-  # Sanity check for well-formed subscribe FilterRequest
-  doAssert(request.subscribe, "invalid subscribe request")
+  if node.wakuFilterClient.isNil():
+    error "cannot register filter subscription to topic", error="waku filter client is nil"
+    return
   
-  info "subscribe content", filter=request
+  let remotePeer = when peer is string: parseRemotePeerInfo(peer) 
+                   else: peer
 
-  var id = generateRequestId(node.rng)
+  info "registering filter subscription to content", pubsubTopic=pubsubTopic, contentTopics=contentTopics, peer=remotePeer
 
-  if not node.wakuFilter.isNil():
-    let
-      pubsubTopic = request.pubsubTopic
-      contentTopics = request.contentFilters.mapIt(it.contentTopic)
+  # Add handler wrapper to store the message when pushed, when relay is disabled and filter enabled
+  # TODO: Move this logic to wakunode2 app
+  let handlerWrapper: FilterPushHandler = proc(pubsubTopic: string, message: WakuMessage) {.raises: [Exception].} =
+      if node.wakuRelay.isNil() and not node.wakuStore.isNil():
+        node.wakuStore.handleMessage(pubSubTopic, message)
 
-    let resSubscription = await node.wakuFilter.subscribe(pubsubTopic, contentTopics)
-    if resSubscription.isOk():
-      id = resSubscription.get()
-    else:
-      # Failed to subscribe
-      error "remote subscription to filter failed", filter = request
-      waku_node_errors.inc(labelValues = ["subscribe_filter_failure"])
+      handler(pubsubTopic, message)
 
-  # Register handler for filter, whether remote subscription succeeded or not
-  node.filters.addContentFilters(id, request.pubSubTopic, request.contentFilters, handler)
-  waku_node_filters.set(node.filters.len.int64)
+  let subRes = await node.wakuFilterClient.subscribe(pubsubTopic, contentTopics, handlerWrapper, peer=remotePeer)
+  if subRes.isOk():
+    info "subscribed to topic", pubsubTopic=pubsubTopic, contentTopics=contentTopics
+  else:
+    error "failed filter subscription", error=subRes.error
+    waku_node_errors.inc(labelValues = ["subscribe_filter_failure"])
 
-proc unsubscribe*(node: WakuNode, request: FilterRequest) {.async, gcsafe.} =
+proc filterUnsubscribe*(node: WakuNode, pubsubTopic: PubsubTopic, contentTopics: ContentTopic|seq[ContentTopic],
```

Comment:
since `ContentTopic` is a subset of `seq[ContentTopic]` can we have just one type `contentTopic:seq[ContentTopic]` instead of `contentTopics: ContentTopic|seq[ContentTopic]` and if its a single one just use `@[topic]`?

Reply:
It is for convenience. The nim compiler compiles one version of the code or another depending on the argument type. Why should we remove it?

Reply:
yes, that's clear. but why doing this? is it some kind of optimization? There are tons of `seq[xxx]` in our code but not in all cases we have `xxx|seq[xxx]` so I guess there is a "why" that I'm missing?